### PR TITLE
esphome 2022.1.2 (new formula)

### PR DESCRIPTION
**Pull request by bobsoppe**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

### CODE REVIEW

File: Formula/esphome.rb [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+
+  desc "Make creating custom firmwares for ESP32/ESP8266 super easy"
+  homepage "https://github.com/esphome/esphome"
+  url "https://files.pythonhosted.org/packages/56/6b/39461f6a4b33155916bbb87fe30334ce5be97b1b8ef4c71a33f73637f89a/esphome-2022.1.2.tar.gz"
+  sha256 "3ce8e499d9d5d8ec4c290f494ef004eb147783e1e700d0a3225e6a68540d5402"
+  license "MIT"
+  revision 1
+
+  depends_on "rust" => :build # for cryptography
+  depends_on "python@3.9"
```
Comment (carlocab) [inline Formula/esphome.rb:12]: Can we use `python@3.10`?

File: Formula/esphome.rb [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+
+  resource "semantic-version" do
+    url "https://files.pythonhosted.org/packages/d4/52/3be868c7ed1f408cb822bc92ce17ffe4e97d11c42caafce0589f05844dd0/semantic_version-2.8.5.tar.gz"
+    sha256 "d2cb2de0558762934679b9a104e82eca7af448c9f4974d1f3eeccff651df8a54"
+  end
+
+  resource "six" do
+    url "https://files.pythonhosted.org/packages/71/39/171f1c67cd00715f190ba0b100d606d440a28c93c7714febeca8b79af85e/six-1.16.0.tar.gz"
+    sha256 "1e61c37477a1626458e36f7b1d82aa5c9b094fa4802892072e49de9c60c4c926"
+  end
```
Comment (carlocab) [inline Formula/esphome.rb:182]: Can we replace this with `depends_on "six"`?

File: Formula/esphome.rb [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+
+  resource "starlette" do
+    url "https://files.pythonhosted.org/packages/11/d3/c346849b8201f0e0339e23850c8db1cebd477e44e8b5212f40e1acbb490c/starlette-0.17.1.tar.gz"
+    sha256 "57eab3cc975a28af62f6faec94d355a410634940f10b30d68d31cb5ec1b44ae8"
+  end
+
+  resource "tabulate" do
+    url "https://files.pythonhosted.org/packages/ae/3d/9d7576d94007eaf3bb685acbaaec66ff4cdeb0b18f1bf1f17edbeebffb0a/tabulate-0.8.9.tar.gz"
+    sha256 "eb1d13f25760052e8931f2ef80aaf6045a6cceb47514db8beab24cded16f13a7"
+  end
```
Comment (carlocab) [inline Formula/esphome.rb:197]: Replace this with `depends_on "python-tabulate"` (if we keep using `python@3.9`)

File: Formula/esphome.rb [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+
+  resource "platformio" do
+    url "https://files.pythonhosted.org/packages/5f/15/607b6732589cbec35ab6d40946d6e038a72fc69893f8f45191bf05a2945d/platformio-5.2.4.tar.gz"
+    sha256 "822ac16e1bcf578b2645466c728f2779e1083ccb4fa21336f4d6539acec53e08"
+  end
+
+  resource "protobuf" do
+    url "https://files.pythonhosted.org/packages/6c/49/f864b9fd6310d9a15ddae5b37d78dff1df0e2e1da476442fee062c6032b2/protobuf-3.19.4.tar.gz"
+    sha256 "9df0c10adf3e83015ced42a9a7bd64e13d06c4cf45c340d2c63020ea04499d0a"
+  end
```
Comment (carlocab) [inline Formula/esphome.rb:137]: Can we do `depends_on "protobuf"` instead?

File: Formula/esphome.rb (hunk 1 above)
Comment (bobsoppe) [reply Formula/esphome.rb]: Changed it to `python@3.10`

File: Formula/esphome.rb (hunk 2 above)
Comment (bobsoppe) [reply Formula/esphome.rb]: Changed the resource to use `depends_on`

File: Formula/esphome.rb (hunk 3 above)
Comment (bobsoppe) [reply Formula/esphome.rb:189]: Updated `python` to `python@3.10`, so not doing this

File: Formula/esphome.rb (hunk 4 above)
Comment (bobsoppe) [reply Formula/esphome.rb]: Changed the resource to use `depends_on`

**Comment by bobsoppe** [comment]
Thank you @carlocab for checking this out!

File: Formula/esphome.rb [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+
+  test do
+    (testpath/"test.yaml").write <<~EOS
+      esphome:
+        name: test
+        platform: ESP8266
+        board: d1
+    EOS
+
+    shell_output("#{bin}/esphome config #{testpath}/test.yaml")
```
Comment (carlocab) [inline Formula/esphome.rb:243]: Can we check output here? Or, if it produces no output:
```suggestion
    assert_empty shell_output("#{bin}/esphome config #{testpath}/test.yaml")
```

**Review by carlocab — COMMENTED** [review]
You'll also need to register `protobuf` and `six` with this formula in `pypi_formula_mappings.json` so that `brew bump-formula-pr` doesn't add those `resource`s again in future version bumps.

File: Formula/esphome.rb [hunk 6]
```diff
@@ -0,0 +1,245 @@
+class Esphome < Formula
+  include Language::Python::Virtualenv
+
+  desc "Make creating custom firmwares for ESP32/ESP8266 super easy"
+  homepage "https://github.com/esphome/esphome"
+  url "https://files.pythonhosted.org/packages/56/6b/39461f6a4b33155916bbb87fe30334ce5be97b1b8ef4c71a33f73637f89a/esphome-2022.1.2.tar.gz"
+  sha256 "3ce8e499d9d5d8ec4c290f494ef004eb147783e1e700d0a3225e6a68540d5402"
+  license "MIT"
+  revision 1
```
Comment (carlocab) [inline Formula/esphome.rb:9]:
```suggestion
```

Comment (bobsoppe) [reply Formula/esphome.rb]: Removed this

File: Formula/esphome.rb (hunk 5 above)
Comment (bobsoppe) [reply Formula/esphome.rb]: Added a more extensive test to check for the output

**Comment by carlocab** [comment]
> > You'll also need to register `protobuf` and `six` with this formula in `pypi_formula_mappings.json` so that `brew bump-formula-pr` doesn't add those `resource`s again in future version bumps.
> 
> I've added those mappings, but I was wondering if we should add those mappings as well for `esptool` and `platformio`, because there are also formulae for those.

`esptool` yes if it works; `platformio` still depends on `python@3.9` so it's not usable as a dependency here. It may be worth adding a comment above the dependencies to test switching the `platformio` resource to a formula dependency when `platformio` is on `python@3.10`.

Thanks for looking at the resources so carefully!

**Comment by bobsoppe** [comment]
> > > You'll also need to register `protobuf` and `six` with this formula in `pypi_formula_mappings.json` so that `brew bump-formula-pr` doesn't add those `resource`s again in future version bumps.
> > 
> > 
> > I've added those mappings, but I was wondering if we should add those mappings as well for `esptool` and `platformio`, because there are also formulae for those.
> 
> `esptool` yes if it works; `platformio` still depends on `python@3.9` so it's not usable as a dependency here. It may be worth adding a comment above the dependencies to test switching the `platformio` resource to a formula dependency when `platformio` is on `python@3.10`.
> 
> Thanks for looking at the resources so carefully!

Wouldn't it be better if we move back to `python@3.9` as it is the default homebrew python version and we can try out `platformio` and `python-tabulate` with it.

**Comment by carlocab** [comment]
Yes, sure, sounds good to me.

**Comment by bobsoppe** [comment]
`esptool` depends on `python@3.10` and using `platformio` was not working. So I did not add those

**Review by carlocab — COMMENTED** [review]
Looks good; thanks!

one last thing: can you squash all your commits into the first one? We need @BrewTestBot to merge this in order to publish the bottles simultaneously with the push to master, but @BrewTestBot doesn’t know how to squash commits that modify non-formula files.

**Comment by bobsoppe** [comment]
I already thought so that it was needed, done now! Thanks for all the help on this one ☺️

**Review by carlocab — APPROVED** [review]
Thanks for your contribution!

**Comment by carlocab** [comment]
We can merge this when CI finishes. Feel free to ping me if I forget to check.

**Comment by bobsoppe** [comment]
@carlocab CI finished, so this could be merged

**Pull request closed without merging**